### PR TITLE
EIP214: STATICCALL

### PR DESCRIPTION
**Pull request by yann300**
Resolves https://github.com/ethereum/cpp-ethereum/issues/3616.

https://github.com/ethereum/EIPs/pull/214

### CODE REVIEW

File: libevm/VM.cpp [hunk 1]
```diff
@@ -214,19 +214,25 @@ void VM::interpretCases()
 		
 		CASE(CREATE)
 		{
+			if (m_schedule->haveStaticCall && m_ext->staticCall)
+				throwBadInstruction();
```
Comment (chfast) [inline libevm/VM.cpp]: Should be `throwDisallowedStateChange()`.

File: libevm/VM.cpp [hunk 2]
```diff
@@ -214,19 +214,25 @@ void VM::interpretCases()
...
-
-			// Pre-homestead
-			if (!m_schedule->haveDelegateCall)
-				throwBadInstruction();
-
+		CASE(STATICCALL)
 		CASE(CALL)
 		CASE(CALLCODE)
 		{
+			// Pre-homestead
```
Comment (chfast) [inline libevm/VM.cpp]: Remove this comment. It does not provided any information any more, as we are so used to hard-forks.

File: libevm/VMCalls.cpp [hunk 3]
```diff
@@ -159,11 +167,13 @@ bool VM::caseCallSetup(CallParameters *callParams, bytesRef& o_output)
 {
 	m_runGas = toInt63(m_schedule->callGas);
 
-	if (m_OP == Instruction::CALL && !m_ext->exists(asAddress(m_SP[1])))
+	callParams->staticCall = (m_OP == Instruction::STATICCALL || m_ext->staticCall);
+
+	if ((m_OP == Instruction::CALL || m_OP == Instruction::STATICCALL) && !m_ext->exists(asAddress(m_SP[1])))
```
Comment (chfast) [inline libevm/VMCalls.cpp]: Should we charge for new account creation in case of STATICCALL? @chriseth @vbuterin?

**Comment by chfast** [comment]
clang has showed an important issue. We missed something...

File: libethereum/ExtVM.h [hunk 4]
```diff
@@ -94,6 +94,7 @@ class ExtVM: public ExtVMFace
 private:
 	State& m_s;  ///< A reference to the base state.
 	SealEngineFace const& m_sealEngine;
+	bool m_staticCall;
```
Comment (chfast) [inline libethereum/ExtVM.h]: This is not needed.

File: libevm/ExtVMFace.h [hunk 5]
```diff
@@ -201,6 +201,7 @@ struct CallParameters
 	u256 apparentValue;
 	u256 gas;
 	bytesConstRef data;
+	bool staticCall;
```
Comment (chfast) [inline libevm/ExtVMFace.h]: Initialize it here too: `bool staticCall = false`.

File: libevm/VM.cpp [hunk 6]
```diff
@@ -214,19 +214,24 @@ void VM::interpretCases()
 		
 		CASE(CREATE)
 		{
+			if (m_schedule->haveStaticCall && m_ext->staticCall)
```
Comment (gumb0) [inline libevm/VM.cpp]: Do we really need a check for `m_schedule->haveStaticCall` here and in other places?
I think it's redundant, it can't be a static call if don't have it 😄

**Comment by winsvega** [comment]
CALLDATA for staticcall is not transfered to the subcall
strange gasUsed count when using different values for memory arguments of STATICCALL

**Comment by chfast** [comment]
Do you want me to help with rebase and cleaning up the history?

File: libethereum/Executive.cpp [hunk 7]
```diff
@@ -240,7 +240,7 @@ bool Executive::execute()
 
 bool Executive::call(Address _receiveAddress, Address _senderAddress, u256 _value, u256 _gasPrice, bytesConstRef _data, u256 _gas)
 {
-	CallParameters params{_senderAddress, _receiveAddress, _receiveAddress, _value, _value, _gas, _data, {}};
+	CallParameters params(_senderAddress, _receiveAddress, _receiveAddress, _value, _value, _gas, _data, {});
```
Comment (chfast) [inline libethereum/Executive.cpp]: This seems to be unnecessary change.

File: libevm/VM.cpp (hunk 6 above)
Comment (chfast) [inline libevm/VM.cpp]: I agree. Keep only `m_ext->staticCall` check.
You can add before `if`: `assert(!staticCall || haveStaticCall)` which checks implication `staticCall => haveStaticCall`.

File: libevm/VM.cpp [hunk 8]
```diff
@@ -265,6 +270,9 @@ void VM::interpretCases()
 
 		CASE(SUICIDE)
 		{
+			if (m_schedule->haveStaticCall && m_ext->staticCall)
```
Comment (chfast) [inline libevm/VM.cpp]: Same here.

File: libevm/VM.cpp [hunk 9]
```diff
@@ -340,6 +348,9 @@ void VM::interpretCases()
 
 		CASE(LOG0)
 		{
+			if (m_schedule->haveStaticCall && m_ext->staticCall)
```
Comment (chfast) [inline libevm/VM.cpp]: Same here.

File: libevm/VMCalls.cpp [hunk 10]
```diff
@@ -159,14 +167,16 @@ bool VM::caseCallSetup(CallParameters *callParams, bytesRef& o_output)
...
```
Comment (chfast) [inline libevm/VMCalls.cpp]: STATICCALL does not have a value argument, so this check does not apply to it.
cc @pirapira.

Comment (pirapira) [inline libevm/VMCalls.cpp]: That's right.

File: test/tools/jsontests/StateTests.cpp [hunk 11]
```diff
@@ -178,6 +178,7 @@ BOOST_AUTO_TEST_CASE(stRevertTest){}
 
 //Metropolis Tests
 BOOST_AUTO_TEST_CASE(stStackTests){}
+BOOST_AUTO_TEST_CASE(stStaticCall){}
```
Comment (chfast) [inline test/tools/jsontests/StateTests.cpp]: @winsvega you forgot to update jsontests.

File: libevm/VMCalls.cpp [hunk 12]
```diff
@@ -182,14 +189,17 @@ bool VM::caseCallSetup(CallParameters *callParams, bytesRef& o_output)
...
-	if (m_OP == Instruction::CALL && !m_ext->exists(asAddress(m_SP[1])))
+	callParams->staticCall = (m_OP == Instruction::STATICCALL || m_ext->staticCall);
+
+	Address destinationAddr = asAddress(m_SP[1]);
+	if (m_OP == Instruction::CALL && !m_ext->exists(destinationAddr))
 		if (m_SP[2] > 0 || m_schedule->zeroValueTransferChargesNewAccountGas())
 			m_runGas += toInt63(m_schedule->callNewAccountGas);
 
-	if (m_OP != Instruction::DELEGATECALL && m_SP[2] > 0)
+	if ((m_OP == Instruction::CALL || m_OP == Instruction::CALLCODE) && m_SP[2] > 0)
```
Comment (chriseth) [inline libevm/VMCalls.cpp]: I hope this won't get us into trouble: We change for value transfer for callcode, although we do not consider it state changing.

**Comment by chriseth** [comment]
Does it make sense to add a check somewhere which verifies that the state root is unchanged after a staticcall? Is that possible from an architectural standpoint?

File: libevm/VMCalls.cpp (hunk 12 above)
Comment (pirapira) [inline libevm/VMCalls.cpp]: What would be the test case for this???

Comment (pirapira) [inline libevm/VMCalls.cpp]: CALLCODE with positive value transfer... probably we already have it...

**Comment by chfast** [comment]
@chriseth, it might be hard because we don't update the state root until the end of a transaction.

File: libevm/VMCalls.cpp (hunk 12 above)
Comment (pirapira) [inline libevm/VMCalls.cpp]: Added a test case in the Metropolis testing table https://docs.google.com/spreadsheets/d/1xat7UI8GtB4ZGVdlK5_XQSHJZaMThi4SrlcL8XMZb5Q/edit#gid=1412823514&range=C21

**Comment by winsvega** [comment]
we could check that state root is not changed in tests.
